packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -444,21 +445,29 @@ export class HardhatNode extends EventEmitter {
   }
 
   public async estimateGas(
-    txParams: TransactionParams,
+    callParams: CallParams,
     blockNumberOrPending: BN | "pending"
   ): Promise<EstimateGasResult> {
-    const tx = await this._getFakeTransaction({
-      ...txParams,
-      gasLimit: this.getBlockGasLimit(),
-    });
+    // We get the CallParams and transform it into a TransactionParams to be
+    // able to run it.
+    const txParams = {
+      ...callParams,
+      nonce: await this._getNonce(
```

Comment:
We were using the latest nonce when estimating gas, and ignoring the block tag. I fixed it.